apm/php: document how to trace long running scripts (#11724)
DataDog/documentation: content/en/tracing/setup_overview/setup/php.md

When [`open_basedir`][12] setting is used, then `/opt/datadog-php` should be added to the list of allowed directories.
When the application runs in a docker container, the path `/proc/self` should also be added to the list of allowed directories.

## Tracing CLI scripts

### Sort-running CLI scripts

> A short-running script runs typically for a few seconds or minutes and the expected behavior is to receive one trace each time the script is executed.

Reviewer: ```suggestion
A short-running script typically runs for a few seconds or minutes and the expected behavior is to receive one trace each time the script is executed.
```

> By default, tracing is disabled for PHP scripts that run from the command line. Opt-in by setting `DD_TRACE_CLI_ENABLED` to `true`.

Reviewer: ```suggestion
By default, tracing is disabled for PHP scripts that run from the command line. Opt in by setting `DD_TRACE_CLI_ENABLED` to `true`.
```

```
$ export DD_TRACE_CLI_ENABLED=true

# Optionally, set the agent host and port if different from localhost and 8126, respectively
$ export DD_AGENT_HOST=agent
$ export DD_TRACE_AGENT_PORT=8126
```

> Assume the following `script.php` that runs a Curl request

Reviewer: ```suggestion
For example, assume the following `script.php` runs a Curl request:
```

```
<?php

sleep(1);

$ch = curl_init('https://httpbin.org/delay/1');
curl_exec($ch);

sleep(1);

```

> Run the script

Reviewer: ```suggestion
Run the script:
```

```
$ php script.php
```

> One trace is generated and sent to the Datadog backend when the script terminates.

Reviewer: ```suggestion
Once run, the trace is generated and sent to the Datadog backend when the script terminates.
```

{{< img src="tracing/setup/php/short-running-cli.jpg" alt="Trace for a short running PHP CLI script" >}}

### Long-running CLI scripts

> A long-running script runs for hours or days. Typically, such scripts execute repetitively a specific task, for example processing new incoming messages or new lines added to a table in a database. The expected behavior is that one trace is generated for each "unit of work", for example the processing of a message.

Reviewer: ```suggestion
A long-running script runs for hours or days. Typically, such scripts repetitively execute a specific task, for example processing new incoming messages or new lines added to a table in a database. The expected behavior is that one trace is generated for each "unit of work", for example the processing of a message.
```

> By default, tracing is disabled for PHP scripts that run from the command line. Opt-in by setting `DD_TRACE_CLI_ENABLED` to `true`.

Reviewer: ```suggestion
By default, tracing is disabled for PHP scripts that run from the command line. Opt in by setting `DD_TRACE_CLI_ENABLED` to `true`.
```

```
$ export DD_TRACE_CLI_ENABLED=true
# With this pair of settings, traces for each "unit of work" is sent as soon as the method execution terminates.
$ export DD_TRACE_GENERATE_ROOT_SPAN=false
$ export DD_TRACE_AUTO_FLUSH_ENABLED=true

# Optionally, set service name, env, etc...
$ export DD_SERVICE=my_service

# Optionally, set the agent host and port if different from localhost and 8126, respectively
$ export DD_AGENT_HOST=agent
$ export DD_TRACE_AGENT_PORT=8126
```

> Assume the following `long_running.php`

Reviewer: ```suggestion
For example, assume the following `long_running.php` script:
```

        curl_exec($ch);
    }
}

/** One of possibly many processing stages, each of which should have a Span */
class ProcessingStage2
{
    public function process(Message $message)
    {
        sleep(1);
    }
}

/** In a real world application, this will read new messages from a source, for example a queue */
function waitForNewMessages()
{
    return [
        new Message($id = (time() + rand(1, 1000)), 'content of a message: ' . $id),
        new Message($id = (time() + rand(1, 1000)), 'content of a message: ' . $id),
        new Message($id = (time() + rand(1, 1000)), 'content of a message: ' . $id),
    ];
}

/** This function is the "unit of work", each execution of it will generate one single trace */
function processMessage(Message $m, array $processors)
{
    foreach ($processors as $processor) {
        $processor->process($m);
        usleep(100000);
    }
}

$processors = [new ProcessingStage1(), new ProcessingStage2()];

/** A loop that runs forever waiting for new messages */
while (true) {
    $messages = waitForNewMessages();
    foreach ($messages as $message) {
        processMessage($message, $processors);
    }
}
```

> Run the script

Reviewer: ```suggestion
Run the script:
```

```
$ php long_running.php
```

> One trace is generated and sent to the Datadog backend every time a new message is processed.

Reviewer: ```suggestion
Once run, one trace is generated and sent to the Datadog backend every time a new message is processed.
```
